### PR TITLE
Add possibility to choose between eye-in-hand or eye-to-hand calibration

### PR DESCRIPTION
**Pull request by SatjaSivcev**
The equivalent of this C++ PR: https://github.com/zivid/zivid-cpp-samples/pull/150

### CODE REVIEW

File: source/applications/advanced/hand_eye_calibration/hand_eye_calibration.py [hunk 1]
```diff
@@ -49,14 +49,35 @@ def _enter_robot_pose(index):
     return robot_pose
 
 
+def _perform_calibration(hand_eye_input):
+    """Calibration type user input.
+
+    Args:
+        calibration_type: Type of calibration, eth (for eye-to-hand) or eih (for eye-in-hand)
```
Comment (robertkt1) [inline source/applications/advanced/hand_eye_calibration/hand_eye_calibration.py:56]: Gives impression that the input argument is calibration type. But agrs is really hand_eye_input?

Comment (SatjaSivcev) [reply source/applications/advanced/hand_eye_calibration/hand_eye_calibration.py]: Good catch. CI also cathed it. Fixed now in https://github.com/zivid/zivid-python-samples/pull/102/commits/07f30aa366d7f015017777f5076e872fcf549e45

File: source/applications/advanced/hand_eye_calibration/hand_eye_calibration.py [hunk 2]
```diff
@@ -83,10 +106,9 @@ def _main():
...
             print(f"Unknown command '{command}'")
 
-    print("Performing Eye-to-Hand calibration")
-    calibration_result = zivid.calibration.calibrate_eye_to_hand(calibration_inputs)
+    calibration_result = _perform_calibration(hand_eye_input)
     code.interact(local=locals())
-    if calibration_result:
+    if calibration_result.valid():
         print("Eye-to-Hand calibration OK")
         print(f"Result:\n{calibration_result}")
```
Comment (robertkt1) [inline source/applications/advanced/hand_eye_calibration/hand_eye_calibration.py:113]: Change to general "Hand eye calibration failed"?

Comment (SatjaSivcev) [reply source/applications/advanced/hand_eye_calibration/hand_eye_calibration.py]: Done in https://github.com/zivid/zivid-python-samples/pull/102/commits/46115ebef3083e25b98c4981a3e9e0238a10add2

**Review by robertkt1 — DISMISSED** [review]
Remember to squash and rebase. Other than that it looks good.

File: source/applications/advanced/hand_eye_calibration/hand_eye_calibration.py [hunk 3]
```diff
@@ -49,14 +49,37 @@ def _enter_robot_pose(index):
...
+        calibration_type = input("Enter type of calibration, eth (for eye-to-hand) or eih (for eye-in-hand):").strip()
+        if calibration_type.lower() == "eth":
+            print("Performing eye-to-hand calibration")
+            hand_eye_output = zivid.calibration.calibrate_eye_to_hand(hand_eye_input)
+            return hand_eye_output
+        if calibration_type.lower() == "eih":
+            print("Performing eye-in-hand calibration")
+            hand_eye_output = zivid.calibration.calibrate_eye_in_hand(hand_eye_input)
+            return hand_eye_output
+        print("Entered unknown method")
```
Comment (dragly) [inline source/applications/advanced/hand_eye_calibration/hand_eye_calibration.py:72]:
```suggestion
        print(f"Unknown method: {calibration_type}")
```

**Review by robertkt1 — COMMENTED** [review]
Should we also rewrite the readme to say perform hand-eye instead of eye-to-hand?

**Comment by SatjaSivcev** [comment]
When I get to input "eih" or "eth" I get strange behavior. Need to investigate more why this happens.

**Comment by SatjaSivcev** [comment]
I've tested this and it works. Ready for review @robertkt1 @torbsorb

File: source/applications/advanced/hand_eye_calibration/hand_eye_calibration.py [hunk 4]
```diff
@@ -83,14 +105,12 @@ def _main():
...
-    code.interact(local=locals())
-    if calibration_result:
-        print("Eye-to-Hand calibration OK")
+    calibration_result = _perform_calibration(hand_eye_input)
+    if calibration_result.valid():
+        print("Hand-eye calibration OK")
         print(f"Result:\n{calibration_result}")
     else:
-        print("Eye-to-Hand calibration FAILED")
+        print("Hand-eye calibration FAILED")
```
Comment (torbsorb) [inline source/applications/advanced/hand_eye_calibration/hand_eye_calibration.py:113]: Can we write Hand-Eye instead?

**Review by torbsorb — APPROVED** [review]
LGTM, with that one minor change

File: source/applications/advanced/hand_eye_calibration/hand_eye_calibration.py (hunk 4 above)
Comment (SatjaSivcev) [reply source/applications/advanced/hand_eye_calibration/hand_eye_calibration.py]: done

**Review by torbsorb — APPROVED** [review]
LGTM

**Comment by SatjaSivcev** [comment]
Ready for next round @torbsorb

File: source/applications/advanced/downsample.py [hunk 5]
```diff
@@ -5,40 +5,10 @@
...
-    visualizer.get_render_option().background_color = (0, 0, 0)
-    visualizer.get_render_option().point_size = 1
-    visualizer.get_render_option().show_coordinate_frame = True
-    visualizer.get_view_control().set_front([0, 0, -1])
-    visualizer.get_view_control().set_up([0, -1, 0])
-
-    visualizer.run()
-    visualizer.destroy_window()
+from sample_utils.paths import get_sample_data_path
+from sample_utils.display import display_pointcloud
```
Comment (torbsorb) [inline source/applications/advanced/downsample.py:11]: This is nice!

File: source/applications/advanced/hand_eye_calibration/ur_hand_eye_calibration/universal_robots_perform_hand_eye_calibration.py [hunk 6]
```diff
@@ -408,7 +408,7 @@ def perform_hand_eye_calibration(mode: str, data_dir: Path):
                 raise RuntimeError(f"Failed to detect feature points from frame {frame_file}")
 
             print(f"Read robot pose from pos{idata:02d}.yaml")
-            with open(pose_file) as file:
+            with open(pose_file, encoding="utf-8") as file:
```
Comment (torbsorb) [inline source/applications/advanced/hand_eye_calibration/ur_hand_eye_calibration/universal_robots_perform_hand_eye_calibration.py:411]: did you consider Path(pose_file).read_text(). I know that there are issues with encoding there as well..

File: source/sample_utils/display.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
+"""
```
Comment (torbsorb) [inline source/sample_utils/display.py:1]: Very nice!

**Review by torbsorb — APPROVED** [review]
LGTM

File: source/applications/advanced/hand_eye_calibration/ur_hand_eye_calibration/universal_robots_perform_hand_eye_calibration.py (hunk 6 above)
Comment (SatjaSivcev) [reply source/applications/advanced/hand_eye_calibration/ur_hand_eye_calibration/universal_robots_perform_hand_eye_calibration.py:411]: No, I haven't. This one is embedded so I prefer it for that reason. I think it is nice to have consistent between samples and some of them will not import Path.

**Pull request merged**